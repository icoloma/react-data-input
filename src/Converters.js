import { isNullOrUndefined, isBlank, passThrough } from "./utils";

const Converters = {
  text: {
    // transform from String representation to object.
    toObject: passThrough,

    // transform from object to String. Null and undefined values are converted to an empty String
    toString: function(value) {
      return isNullOrUndefined(value) ? "" : value;
    }
  },

  url: {
    // transform from String representation to object.
    toObject: function(value) {
      return isBlank(value) ? undefined : value;
    },

    toString: function(value) {
      return Converters.text.toString(value);
    }
  },

  number: {
    toObject: function(value, attrs) {
      return isBlank(value)
        ? undefined
        : attrs.step == 1
        ? parseInt(value)
        : parseFloat(value);
    },

    toString: function(value) {
      return isNullOrUndefined(value) ? "" : "" + value;
    }
  },

  checkbox: {
    toObject: passThrough,
    toString: passThrough
  }
};

export default Converters;
