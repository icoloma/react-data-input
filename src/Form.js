import React from "react";
import PropTypes from "prop-types";

// Form is the (optional) container of BoundComponent instances.
export default class Form extends React.Component {
  constructor() {
    super();
    this.onSubmit = this.onSubmit.bind(this);
    this.validationComponents = [];
    this.ref = React.createRef();
  }

  getChildContext() {
    return {
      form: this
    };
  }

  addValidationComponent(c) {
    this.validationComponents.push(c);
  }

  removeValidationComponent(c) {
    const vc = this.validationComponents;
    const index = vc.indexOf(c);
    index !== -1 && vc.splice(index, 1);
  }

  // validate all child components
  // @return a Promise that resolves to true if validation passes for all fields, false otherwise
  validate() {
    const promises = this.validationComponents.map(c => c.validate());
    return Promise.all(promises).then(errorMessages => {
      return errorMessages.every(errorMessage => !errorMessage);
    });
  }

  // if all validations pass, invokes the onSubmit callback
  // @return a Promise to be invoked after the validation is done
  onSubmit(e) {
    e.preventDefault();
    return this.validate().then(results => {
      if (results) {
        this.props.onSubmit(e);
      } else {
        const firstInvalidElement = this.ref.current.querySelector(
          "[aria-invalid=true]"
        );
        firstInvalidElement && firstInvalidElement.focus();
      }
    });
  }

  render() {
    const { state, ...props } = this.props;
    return (
      <form {...props} onSubmit={this.onSubmit} noValidate ref={this.ref}>
        {this.props.children}
      </form>
    );
  }
}

Form.propTypes = {
  // the callback that will be invoked on submit, if validation passes
  onSubmit: PropTypes.func.isRequired,

  // the state that will be injected into nested elements
  state: PropTypes.object
};

Form.childContextTypes = {
  form: PropTypes.object
};
