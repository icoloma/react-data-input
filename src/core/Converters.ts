import { isNullOrUndefined, isBlank } from "./utils";
import { InputHTMLAttributes } from "react";

function emptyToUndef({ value }): string | undefined {
  return isBlank(value) ? undefined : value;
}

function nullOrUndefToEmpty(value: string | null | undefined): string {
  return isNullOrUndefined(value) ? "" : value;
}

export interface Converter<V> {
  /** Convert from the HTML attributes into a JavaScript value */
  fromValue(props: InputHTMLAttributes<HTMLInputElement>): V;

  /** Convert from JavaScript value into the `value` HTML attribute */
  toValue(input: V): string;
}

interface ConvertersType {
  text: Converter<string>;
  url: Converter<string>;
  number: Converter<number>;
  date: Converter<string>;
  time: Converter<string>;
}

export const Converters: ConvertersType = {
  text: {
    fromValue({ value }) {
      return value as string;
    },
    toValue: nullOrUndefToEmpty,
  },

  url: {
    fromValue: emptyToUndef,
    toValue: nullOrUndefToEmpty,
  },

  number: {
    fromValue({ value, step }) {
      const input = value as string;
      return isBlank(input)
        ? undefined
        : step == 1
        ? parseInt(input)
        : parseFloat(input);
    },

    toValue(input) {
      return isNullOrUndefined(input) ? "" : "" + input;
    },
  },

  date: {
    fromValue: emptyToUndef,
    toValue: nullOrUndefToEmpty,
  },

  time: {
    fromValue: emptyToUndef,
    toValue: nullOrUndefToEmpty,
  },
};
