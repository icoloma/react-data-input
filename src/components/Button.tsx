import { ButtonHTMLAttributes, useContext } from "react";
import React from "react";
import { FormContext } from "./Form";

/**
 * A submit button that can display an "in progress" css style. Its use is optional.
 * Use: <Button className="mybutton">Save</Button>
 */
export function Button({
  className = "",
  children,
  ...props
}: ButtonHTMLAttributes<HTMLButtonElement>) {
  const { submitting } = useContext(FormContext);
  return (
    <button
      className={`${submitting ? "in-progress " : ""}${className}`}
      {...props}
    >
      {children}
    </button>
  );
}
