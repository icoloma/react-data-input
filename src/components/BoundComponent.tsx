import React, { useContext, ChangeEvent, useEffect, useState } from "react";
import { Converters, Converter } from "../core/Converters";
import { InputProps } from "../core/ValidationActions";
import { FormContext, FormContextContent } from "./Form";

export interface BoundComponentProps extends InputProps {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";

  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<any>;
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent({
  elementName,
  converter,
  name,
  id,
  onChange: originalOnChange,
  children,
  ...props
}: BoundComponentProps) {
  const type = props.type;
  if (process.env.NODE_ENV != "production") {
    if (type === "radio" && !props.defaultValue) {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
    if (typeof props.value !== "undefined") {
      throw new Error(`${name} should use defaultValue instead of value`);
    }
  }

  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const defaultValue =
    type === "checkbox"
      ? undefined
      : type === "radio"
      ? (props.defaultValue as string)
      : converter.objectToHtml(formContext.getValue(name));
  const defaultChecked =
    type === "checkbox"
      ? formContext.getValue(name)
      : type === "radio"
      ? formContext.getValue(name) == defaultValue
      : undefined;

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter.htmlToObject({ value: element.value, ...props });
    formContext.setValue(name, objectValue);
    originalOnChange && originalOnChange(e);
  }

  useEffect(() => {
    formContext.addValidator(name, props);
    // TODO: should we do something to unmount?
  }, []);

  const error = formContext.errors[name];
  const errorMessageId = error ? `${id}_error` : undefined;

  // ARIA attributes to point to the error message
  const adb = props["aria-describedby"];
  const ariaProps = !errorMessageId
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          defaultChecked,
          defaultValue,
          ...ariaProps,
          ...props,
        },
        children
      )}
      {error ? (
        <div className="input-error" id={errorMessageId} role="alert">
          {error}
        </div>
      ) : undefined}
    </>
  );
}
