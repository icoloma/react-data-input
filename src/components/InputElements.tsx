import React, { useContext, ChangeEvent, useEffect, useState } from "react";
import { Converters, Converter } from "../core/Converters";
import { InputProps } from "../core/ValidationActions";
import { FormContext, FormContextContent } from "./Form";

export interface BoundComponentProps extends InputProps {
  /** type of input component to use */
  elementName: "input" | "select" | "textarea";

  /** propertyName. Will also be used as id if none is specified */
  name: string;

  /** Optional: the converter to use. If not set, the default for the input type will be used */
  converter?: Converter<any>;
}

/**
 * An input/select/textarea component that will bind automatically
 * with the corresponding object propety in the containing `Form`
 */
export function BoundComponent({
  elementName,
  converter,
  name,
  id,
  onChange: originalOnChange,
  children,
  ...props
}: BoundComponentProps) {
  const type = props.type;
  if (process.env.NODE_ENV != "production") {
    if (type === "radio" && !props.defaultValue) {
      throw new Error(`Radio button ${name} requires a defaultValue`);
    }
    if (typeof props.value !== "undefined") {
      throw new Error(`${name} should use defaultValue instead of value`);
    }
  }

  id = id || type === "radio" ? `${name}_${props.defaultValue}` : name;
  converter = converter || Converters[type] || Converters.text;
  const formContext = useContext<FormContextContent>(FormContext);
  const defaultValue =
    type === "checkbox"
      ? undefined
      : type === "radio"
      ? (props.defaultValue as string)
      : converter.objectToHtml(formContext.getValue(name));
  const defaultChecked =
    type === "checkbox"
      ? formContext.getValue(name)
      : type === "radio"
      ? formContext.getValue(name) == defaultValue
      : undefined;

  function onChange(e: ChangeEvent<HTMLInputElement>) {
    const element = e.target;
    const objectValue =
      type === "checkbox"
        ? element.checked
        : converter.htmlToObject({ value: element.value, ...props });
    formContext.setValue(name, objectValue);
    originalOnChange && originalOnChange(e);
  }

  useEffect(() => {
    formContext.addValidator(name, props);
    // TODO: should we do something to unmount?
  }, []);

  const error = formContext.errors[name];

  // ARIA attributes to point to the error message
  const errorMessageId = error ? `${id}_error` : undefined;
  const adb = props["aria-describedby"];
  const ariaProps = !error
    ? {}
    : {
        "aria-invalid": true,
        "aria-describedby": adb ? `${errorMessageId} ${adb}` : errorMessageId,
      };

  return (
    <>
      {React.createElement(
        elementName,
        {
          name,
          id: id,
          onChange,
          defaultChecked,
          defaultValue,
          ...ariaProps,
          ...props,
        },
        children
      )}
      {error ? (
        <div className="input-error" id={errorMessageId} role="alert">
          {error}
        </div>
      ) : undefined}
    </>
  );
}

type WrappedBoundComponentProps = Omit<BoundComponentProps, "elementName">;

interface ExtendedInputProps extends WrappedBoundComponentProps {
  type: "text" | "url" | "number" | "checkbox" | "date" | "time" | "radio";
}

export function Input(props: ExtendedInputProps) {
  return <BoundComponent elementName="input" {...props} />;
}

export function Select(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="textarea" {...props} />;
}
