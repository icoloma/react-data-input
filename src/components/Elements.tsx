import React from "react";
import { BoundComponent, BoundComponentProps } from "./BoundComponent";

type WrappedBoundComponentProps = Omit<BoundComponentProps, "elementName">;

interface ExtendedInputProps extends WrappedBoundComponentProps {
  type: "text" | "url" | "number" | "checkbox" | "date" | "time" | "radio";
}

export function Input(props: ExtendedInputProps) {
  return <BoundComponent elementName="input" {...props} />;
}

export function Select(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="select" {...props} />;
}

export function TextArea(props: WrappedBoundComponentProps) {
  return <BoundComponent elementName="textarea" {...props} />;
}
