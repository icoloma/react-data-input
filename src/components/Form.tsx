import React, {
  FormHTMLAttributes,
  ReactEventHandler,
  createContext,
  useState,
  useContext,
  useEffect,
} from "react";
import {
  filterActionsForProps,
  ValidationResult,
} from "../core/ValidationActions";
import { I18nContext } from "./I18nContext";
import { Messages } from "../core/Messages";
import { Validators, Validator } from "../core/Validators";
import { BoundComponentProps } from "./InputElements";

interface FormProps extends FormHTMLAttributes<HTMLFormElement> {
  /** the object being edited  */
  initialValues: any;

  /** the method to invoke when submitting, after passing all validations */
  onSubmit: (values: any) => Promise<any>;
}

/**
 * The list of all validation errors. The key is the propertyName,
 * the value is the error message
 */
export interface ValidationErrors {
  [propertyName: string]: string;
}

/**
 * Context to share with child BoundComponents
 */
export interface FormContextContent {
  /**
   * All error messages, indexed by property name
   */
  errors: ValidationErrors;

  /**
   * True if this form is submitting the value
   */
  submitting: boolean;

  /**
   * Add a Validator to the list of fields validated by this form
   */
  addValidator(props: BoundComponentProps): void;

  /**
   * Triggered by a change in the form field
   */
  setValue(propertyName: string, value: any): void;

  /**
   * Get value associated to a BoundComponent
   */
  getValue(propertyName: string): any;
}
function undefinedFormContext(): never {
  throw new Error("A <Form> context is required");
}
export const FormContext = createContext<FormContextContent>({
  addValidator: undefinedFormContext,
  setValue: undefinedFormContext,
  getValue: undefinedFormContext,
  errors: {},
  submitting: false,
});

/**
 * Form is the required container  of BoundComponent instances.
 */
export function Form({
  initialValues,
  onSubmit,
  children,
  ...props
}: FormProps) {
  // errors to display, sorted by component key
  const [errors, setErrors] = useState<ValidationErrors>({});

  // object with the data introduced in this form
  const [values, setValues] = useState(initialValues);

  // the list of validators to use for this object
  const [validators, setValidators] = useState<Validators>({});

  // the submission status
  const [submitting, setSubmitting] = useState(false);

  // should focus on the first component with error
  const [focusFirstError, setFocusFirstError] = useState(false);

  // validation rules to apply
  const entries = useContext(I18nContext);
  const messages = new Messages(entries);

  const formContextValue: FormContextContent = {
    errors,
    submitting,
    addValidator: function (props) {
      validators[props.name] = new Validator(props);
      setValidators(validators);
    },

    setValue: function (propertyName, value) {
      values[propertyName] = value;
      if (errors[propertyName]) {
        delete errors[propertyName];
        setErrors({ ...errors });
      }
    },

    getValue(propertyName) {
      return values[propertyName];
    },
  };

  async function validate(): Promise<ValidationErrors> {
    const newErrors: ValidationErrors = {};
    await Promise.all(
      Object.entries(validators).map(async ([propertyName, validator]) => {
        const error: ValidationResult = await validator.validate(
          values[propertyName]
        );
        if (!!error) {
          newErrors[propertyName] = messages.get(error, validator.props);
        }
      })
    );
    setErrors(newErrors);
    if (Object.keys(newErrors).length > 0) {
      setFocusFirstError(true);
    }
    return newErrors;
  }

  const onSubmitHandler: ReactEventHandler = async function (e): Promise<any> {
    e.preventDefault();
    const errors = await validate();
    if (!Object.keys(errors).length) {
      setSubmitting(true);
      return onSubmit(values).finally(() => setSubmitting(false));
    } else {
      return Promise.reject(errors);
    }
  };

  // todo: only do this when fully validating, not by setErrors()

  useEffect(() => {
    if (focusFirstError) {
      const firstInvalidElement = document.body.querySelector<HTMLInputElement>(
        "[aria-invalid=true]"
      );
      firstInvalidElement && firstInvalidElement.focus();
      setFocusFirstError(false);
    }
  }, [errors]);

  return (
    <form {...props} onSubmit={onSubmitHandler} noValidate>
      <FormContext.Provider value={formContextValue}>
        {children}
      </FormContext.Provider>
    </form>
  );
}

/**
 * A fieldset that will disable all input controls while submitting
 */
export function FieldSet({ children, ...props }) {
  const formContext = useContext(FormContext);
  return (
    <fieldset disabled={formContext.submitting} {...props}>
      {children}
    </fieldset>
  );
}
