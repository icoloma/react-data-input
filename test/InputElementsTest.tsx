import React, {
  ReactElement,
  ChangeEvent,
  HTMLAttributes,
  InputHTMLAttributes,
} from "react";
import renderer, {
  ReactTestRenderer,
  act,
  ReactTestInstance,
} from "react-test-renderer";
import { Input } from "../src/components/InputElements";
import { Form, ValidationErrors } from "../src/components/Form";
import { Converters } from "../src/core/Converters";

describe("Input", function () {
  let onSubmit: () => Promise<string>;

  let form: ReactTestRenderer;

  function mount(input: ReactElement, initialValues: any) {
    onSubmit = jest.fn(async () => "foo");
    // act() to wait for the useEffect() calls on initial render
    act(() => {
      form = renderer.create(
        <Form onSubmit={onSubmit} initialValues={initialValues}>
          {input}
        </Form>
      );
    });
  }
  /*
  function waitForElement(props: any): Promise<ReactTestInstance[]> {
    return new Promise<ReactTestInstance[]>((resolve, reject) => {
      let tries = 0;
      function check() {
        const elements = form.root.findAll((el) =>
          Object.entries(props).every(
            ([propertyName, value]) => el[propertyName] == value
          )
        );
        if (elements.length) {
          resolve(elements);
        }
        if (tries++ < 5) {
          setTimeout(check, 10);
        } else {
          reject(
            new Error(
              `Timeout waiting for element with props: ${JSON.stringify(
                props
              )}.\nCurrent tree: ${JSON.stringify(form.toJSON(), null, "  ")}`
            )
          );
        }
      }
      setTimeout(check, 10);
    });
  }
*/
  async function triggerChange({
    value,
    checked,
  }: {
    value: string;
    checked?: boolean;
  }) {
    return act(() => {
      const inputElement = form.root.find((el) => el.type == "input");
      expect(inputElement.type).toBe("input");
      inputElement.props.onChange({
        target: {
          value,
          checked,
        },
      });
    });
  }

  async function triggerSubmit(expectErrors?: ValidationErrors) {
    return act(async () => {
      const formElement = form.root.find((el) => el.type == "form");
      try {
        await formElement.props.onSubmit({ preventDefault() {} });
        expect(expectErrors).toBeUndefined();
      } catch (errors) {
        expect(errors).toMatchObject(expectErrors);
      }
    });
  }

  it("should render correctly with type=text", () => {
    mount(<Input type="text" className="foo" required name="name" />, {
      name: "foobar",
    });
    // input has class 'foo'
    // input is required
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should use checked instead of value with type=checkbox", () => {
    mount(<Input type="checkbox" name="subscribed" />, { subscribed: true });
    // input[checked] is set
    // input[value] is not set
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render type=radio", () => {
    mount(
      <>
        <Input type="radio" name="abtesting" defaultValue="a" />
        <Input type="radio" name="abtesting" defaultValue="b" />
      </>,
      { abtesting: "a" }
    );
    // input[checked] is set
    // different id calculated for each input
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render type=radio", function () {
    mount(
      <>
        <Input type="radio" name="gender" defaultValue="male" />,
        <Input type="radio" name="gender" defaultValue="female" />
      </>,
      { gender: "male" }
    );
    // male selected, female unselected
    expect(form.toJSON()).toMatchSnapshot(
      "radio: male selected, female unselected"
    );

    // trigger change in input2
    const input2 = form.root.findAll((el) => el.type == "input")[1];
    input2.props.onChange({
      target: { value: "female", checked: true },
    });
    expect(form.toJSON()).toMatchSnapshot(
      "radio: female selected, male unselected"
    );
  });

  it("should not propagate specific properties, like converter, to the HTML5 element", () => {
    mount(
      <Input
        type="text"
        className="foo"
        name="name"
        converter={Converters.text}
      />,
      {}
    );
    // state attribute should not be propagated to nested <input>
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render a validation error", async function () {
    mount(<Input type="text" required name="name" />, {});
    await triggerSubmit({ name: "Please fill out this field" });
    // error message was rendered
    // error message is accessible: role=alert, aria-invalid, aria-describedby
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should update number (also: onChange listener)", async () => {
    const values = { age: 20, color: "red" };
    const callback = jest.fn();
    mount(
      <Input type="number" name="age" step="1" onChange={callback} />,
      values
    );
    await triggerChange({ value: "32.2" });
    expect(values).toMatchObject({
      age: 32,
      color: "red",
    });
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("required and optional checkboxes (also: set id attribute)", async function () {
    mount(
      <Input
        type="checkbox"
        name="subscribed"
        id="foobarbaz"
        required={false}
      />,
      {
        subscribed: false,
      }
    );
    await triggerSubmit();

    mount(
      <Input
        type="checkbox"
        name="subscribed"
        id="foobarbaz"
        required={true}
      />,
      {
        subscribed: false,
      }
    );
    await triggerSubmit({ subscribed: "Please fill out this field" });
    expect(form.toJSON()).toMatchSnapshot();
  });

  /*
  function initAsyncValidator(state) {
    const validator = (value) =>
      new Promise((resolve) => {
        resolve(value === "abc" ? undefined : "Validation failed");
      });
    mount(
      <Input type="text" name="foo" validator={validator} />,
      state
    );
    return form.getInstance();
  }

  it("custom validator with Promise expected to fail", function () {
    const state = { foo: "xyz" };
    initAsyncValidator(state);
    return form.validate().then((result) => {
      expect(result).toBe(false);
    });
  });

  it("custom validator with Promise expected to pass", function () {
    const state = { foo: "abc" };
    initAsyncValidator(state);
    return form.validate().then((result) => {
      expect(result).toBe(true);
    });
  });

  */
});
