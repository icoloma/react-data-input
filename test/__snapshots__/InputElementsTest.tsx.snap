// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input custom async validation fails (also: onSubmit not called) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="foo_error"
    aria-invalid={true}
    defaultValue="xyz"
    id="foo"
    name="foo"
    onChange={[Function]}
    type="text"
  />
  <div
    className="input-error"
    id="foo_error"
    role="alert"
  >
    Please match the requested format
  </div>
</form>
`;

exports[`Input custom async validation passes (also: onSubmit called) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultValue="abc"
    id="foo"
    name="foo"
    onChange={[Function]}
    type="text"
  />
</form>
`;

exports[`Input required and optional checkboxes (also: set id attribute) 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="foobarbaz_error"
    aria-invalid={true}
    defaultChecked={false}
    id="foobarbaz"
    name="subscribed"
    onChange={[Function]}
    required={true}
    type="checkbox"
  />
  <div
    className="input-error"
    id="foobarbaz_error"
    role="alert"
  >
    Please fill out this field
  </div>
</form>
`;

exports[`Input should not propagate specific properties, like converter, to the HTML5 element 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    className="foo"
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    type="text"
  />
</form>
`;

exports[`Input should render a validation error 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    aria-describedby="name_error"
    aria-invalid={true}
    defaultValue=""
    id="name"
    name="name"
    onChange={[Function]}
    required={true}
    type="text"
  />
  <div
    className="input-error"
    id="name_error"
    role="alert"
  >
    Please fill out this field
  </div>
</form>
`;

exports[`Input should render correctly with type=text 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    className="foo"
    defaultValue="foobar"
    id="name"
    name="name"
    onChange={[Function]}
    required={true}
    type="text"
  />
</form>
`;

exports[`Input should render type=radio 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultChecked={true}
    defaultValue="a"
    id="abtesting_a"
    name="abtesting"
    onChange={[Function]}
    type="radio"
  />
  <input
    defaultChecked={false}
    defaultValue="b"
    id="abtesting_b"
    name="abtesting"
    onChange={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should render type=radio: radio: female selected, male unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultChecked={true}
    defaultValue="male"
    id="gender_male"
    name="gender"
    onChange={[Function]}
    type="radio"
  />
  ,
  <input
    defaultChecked={false}
    defaultValue="female"
    id="gender_female"
    name="gender"
    onChange={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should render type=radio: radio: male selected, female unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultChecked={true}
    defaultValue="male"
    id="gender_male"
    name="gender"
    onChange={[Function]}
    type="radio"
  />
  ,
  <input
    defaultChecked={false}
    defaultValue="female"
    id="gender_female"
    name="gender"
    onChange={[Function]}
    type="radio"
  />
</form>
`;

exports[`Input should update nested properties 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultValue=""
    id="from-address"
    name="from.address"
    onChange={[Function]}
    required={true}
    type="text"
  />
  ,
  <input
    defaultValue="barbaz"
    id="to-address"
    name="to.address"
    onChange={[Function]}
    type="text"
  />
  ,
</form>
`;

exports[`Input should use checked instead of value with type=checkbox 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    defaultChecked={true}
    id="subscribed"
    name="subscribed"
    onChange={[Function]}
    type="checkbox"
  />
</form>
`;
